crates/wit-component/src/decoding.rs

```diff
@@ -1,190 +1,196 @@
+                    for export in s {
+                        let export = export?;
+                        let prev = exports.insert(export.name, export);
+                        assert!(prev.is_none());
+                    }
+                }
+                _ => {}
             }
         }
+        Ok(Self {
+            types: types.unwrap(),
+            imports,
+            exports,
+        })
     }
 }
 
 /// Represents an interface decoder for WebAssembly components.
-pub struct InterfaceDecoder<'a> {
+struct InterfaceDecoder<'a> {
     info: &'a ComponentInfo<'a>,
     interface: Interface,
-    type_map: IndexMap<types::TypeId, Type>,
-    name_map: IndexMap<types::TypeId, &'a str>,
+
+    // Note that the hash keys in these maps are `&types::Type` where we're
+    // hashing the memory address of the pointer itself. The purpose here is to
+    // ensure that two `TypeId` entries which come from two different index
+    // spaces which point to the same type can name the same type, so the hash
+    // key is the result after `TypeId` lookup.
+    type_map: IndexMap<PtrHash<'a, types::Type>, Type>,
+    name_map: IndexMap<PtrHash<'a, types::Type>, &'a str>,
+}
+
+/// Parsed representation of interfaces found within a component.
+///
+/// This is more-or-less a "world" and will likely be replaced one day with a
+/// `wit-parser` representation of a world.
+#[derive(Default)]
+pub struct ComponentInterfaces<'a> {
+    /// The "default export" which is the interface directly exported from the
+    /// component at the first level.
+    pub default: Option<Interface>,
+    /// Imported interfaces, keyed by name, to the component.
+    pub imports: IndexMap<&'a str, Interface>,
+    /// Exported interfaces, keyed by name, to the component.
+    pub exports: IndexMap<&'a str, Interface>,
+}
+
+/// Decode the interfaces imported and exported by a component.
+///
+/// This function takes a binary component as input and will infer the
+/// `Interface` representation of its imports and exports. More-or-less this
+/// will infer the "world" from a binary component. The binary component at this
+/// time is either a "types only" component produced by `wit-component` or an
+/// actual output of `wit-component`.
+///
+/// The returned interfaces represent the description of imports and exports
+/// from the component.
+///
+/// This can fail if the input component is invalid or otherwise isn't of the
+/// expected shape. At this time not all component shapes are supported here.
+pub fn decode_interface_component(bytes: &[u8]) -> Result<ComponentInterfaces<'_>> {
+    let info = ComponentInfo::new(bytes)?;
+    let mut imports = IndexMap::new();
+    let mut exports = IndexMap::new();
+
+    for (name, ty) in info.imports.iter() {
+        // Imports right now are only supported if they're an import of an
+        // instance and an instance is self-describing in the interface that it
+        // represents.
```

Comment:
Can you clarify what self-describing means here - I don't think any of the instance imports presently available in the  binary format are not self-describing? or is this about future *-imports?

Reply:
Oh I just meant that the type of the import is sort of the serialization of the `Interface` type, I reworded a bit to capture that.